{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Expression Programming"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports\n",
    "External libraries used:\n",
    "* AnyTree (packaged locally, no installation required)\n",
    "* MatPlotLib\n",
    "* Numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from copy import deepcopy\n",
    "from lib.anytree.node import Node\n",
    "from lib.anytree.render import RenderTree\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from random import random, randint, shuffle\n",
    "from warnings import warn\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Chromosome Class - Container for a list of genes (strings)\n",
    "Contains methods to...\n",
    "* Build an expression tree from a gene\n",
    "* Evaluate an expression tree given fitness cases and terminal values\n",
    "* Display the chromosome's expression tree to the console\n",
    "* Link chromosomes\n",
    "* Calculate fitness (several ways)\n",
    "* Generate random genes/chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "class Chromosome:\n",
    "\n",
    "    # Functions and Terminals are shared by all chromosomes\n",
    "    functions = dict()\n",
    "    terminals = list()\n",
    "    constants = dict()\n",
    "    ephemeral_random_constants_range = (-1, 1)\n",
    "    linking_function = None\n",
    "\n",
    "    # length of head of chromosome\n",
    "    num_genes = 3\n",
    "    head_length = 6\n",
    "    length = 39\n",
    "\n",
    "    # list of real-valued tuples of the form (x, f(x))\n",
    "    fitness_cases = []\n",
    "    max_fitness = None\n",
    "\n",
    "\n",
    "    def __init__(self, genes: list):\n",
    "\n",
    "        # do not let chromosomes be defined without first defining their functions, terminals, and head length\n",
    "        if not Chromosome.functions:\n",
    "            raise ValueError(\"Chromosome class has no functions associated with it.\")\n",
    "        if len(Chromosome.terminals) == 0:\n",
    "            raise ValueError(\"Chromosome class has no terminals associated with it.\")\n",
    "        if Chromosome.length is None:\n",
    "            raise ValueError(\"Chromosome class has no length defined.\")\n",
    "        if Chromosome.head_length is None:\n",
    "            raise ValueError(\"Chromosome class has no head length defined.\")\n",
    "        if Chromosome.linking_function is None and len(genes) > 1:\n",
    "            raise ValueError(\"Multigenic chromosome defined with no linking function.\")\n",
    "        if len(genes) != Chromosome.num_genes:\n",
    "            raise ValueError(\"Number of genes does not match excpected value in class level variable.\")\n",
    "        if \"?\" in Chromosome.terminals and Chromosome.ephemeral_random_constants_range is None:\n",
    "            raise ValueError(\"Must define ephemeral random constants range if using ephemeral random constants.\")\n",
    "\n",
    "        # initialize chromosomes\n",
    "        self.genes = genes\n",
    "        self.trees = []\n",
    "        self._values_ = {}\n",
    "        self._fitness_ = None\n",
    "        self.ephemeral_random_constants = list(np.random.uniform(*Chromosome.ephemeral_random_constants_range, size=Chromosome.length))\n",
    "\n",
    "\n",
    "    # TODO - put informative error message when terminal_values doesn't have enough entries\n",
    "    def evaluate(self, terminal_values: dict) -> float:\n",
    "        \"\"\"\n",
    "        Returns the result of evaluating the given chromosome for specified fitness cases.\n",
    "\n",
    "        :param terminal_values: dictionary mapping all present terminal symbols to real values\n",
    "        :return: real valued result of evaluating the chromosome\n",
    "        \"\"\"\n",
    "\n",
    "        # memoize value in case the chromosome was already evaluated\n",
    "        value_fingerprint = tuple(sorted(terminal_values.items()))\n",
    "        if value_fingerprint in self._values_:\n",
    "            return self._values_[value_fingerprint]\n",
    "\n",
    "        # build expression trees for each gene if not already built\n",
    "        if len(self.trees) == 0:\n",
    "            self.trees = [Chromosome.build_tree(gene) for gene in self.genes]\n",
    "\n",
    "        # link expression trees if the chromosome is multigenic, otherwise use first tree\n",
    "        if self.num_genes > 1:\n",
    "            expression_tree = Chromosome.link(*self.trees)\n",
    "        else:\n",
    "            expression_tree = self.trees[0]\n",
    "\n",
    "        erc_index = 0\n",
    "\n",
    "        # recursive inorder tree traversal\n",
    "        def inorder(start: Node) -> float:\n",
    "            nonlocal terminal_values, erc_index\n",
    "            if start.name in Chromosome.terminals:\n",
    "                if start.name == \"?\":\n",
    "                    erc_index += 1\n",
    "                    return self.ephemeral_random_constants[erc_index - 1]\n",
    "                if start.name in Chromosome.constants:\n",
    "                    return Chromosome.constants[start.name]\n",
    "                return int(start.name) if start.name.isdigit() else terminal_values[start.name]\n",
    "            if start.name in Chromosome.functions:\n",
    "                return Chromosome.functions[start.name][\"f\"](*[inorder(node) for node in start.children])\n",
    "\n",
    "        try:\n",
    "            self._values_[value_fingerprint] = inorder(expression_tree)\n",
    "            if isinstance(self._values_[value_fingerprint], np.complex):\n",
    "                raise TypeError\n",
    "        # ZeroDivisionError if tree does something like x/(y-y), TypeError if the takes square root of a negative.\n",
    "        except (ZeroDivisionError, TypeError):\n",
    "            self._values_[value_fingerprint] = np.nan\n",
    "\n",
    "        # noinspection PyTypeChecker\n",
    "        return self._values_[value_fingerprint]\n",
    "\n",
    "\n",
    "    def fitness(self) -> float:\n",
    "        \"\"\"\n",
    "        Getter for fitness property to make sure we aren't grabbing uncalculated fitnesses\n",
    "\n",
    "        :return: fitness of chromosome, or raise a warning and return 0 if fitness hasn't been calculated\n",
    "        \"\"\"\n",
    "        if self._fitness_ is not None:\n",
    "            return self._fitness_\n",
    "        warn(\"Fitness of chromosome has not been properly calculated. Returning 0.\")\n",
    "        return 0\n",
    "\n",
    "\n",
    "    def print_tree(self) -> None:\n",
    "        \"\"\"\n",
    "        Use AnyTree to display a Chromosome's expression tree(s)\n",
    "\n",
    "        :return: void\n",
    "        \"\"\"\n",
    "        for t in range(len(self.trees)):\n",
    "            print(\"Tree %d\" % t)\n",
    "            for pre, _, node in RenderTree(self.trees[t]):\n",
    "                print(\"\\t%s%s\" % (pre, node.name))\n",
    "        print(self.ephemeral_random_constants)\n",
    "\n",
    "    def plot_solution(self, objective_function, x_min: float, x_max: float,\n",
    "                      avg_fitnesses: list, best_fitnesses: list, variable_name: str) -> None:\n",
    "\n",
    "        \"\"\"\n",
    "        Mostly unused, handy for plotting symbolic regression results though\n",
    "\n",
    "        :param objective_function: ground truth function to plot\n",
    "        :param x_min: minimum x value of plot\n",
    "        :param x_max: maximum x value of plot\n",
    "        :param avg_fitnesses: list of average fitness values by generation\n",
    "        :param best_fitnesses: list of best fitness values by generation\n",
    "        :return: void\n",
    "        \"\"\"\n",
    "\n",
    "        if objective_function is not None:\n",
    "            # set up subplots\n",
    "            plt.subplots(1, 2, figsize=(16, 8))\n",
    "\n",
    "            # Objective function vs Discovered function plot\n",
    "            xs = np.linspace(x_min, x_max, 100)\n",
    "            plt.subplot(1, 2, 1)\n",
    "            plt.title(\"Discovered function vs. Objective function\")\n",
    "            plt.plot(xs, [objective_function(x) for x in xs],\n",
    "                     linewidth=2, linestyle='dashed', color='black', label=\"Objective\")\n",
    "            plt.plot(xs, [self.evaluate({variable_name: x}) for x in xs],\n",
    "                     linewidth=2, color='blue', label=\"Discovered\")\n",
    "            plt.legend(loc=\"upper left\")\n",
    "\n",
    "            # Fitness over time plot\n",
    "            plt.subplot(1, 2, 2)\n",
    "\n",
    "            plt.title(\"Fitness by Generation\")\n",
    "            plt.plot(range(len(avg_fitnesses)), avg_fitnesses, label=\"Average\")\n",
    "            plt.plot(range(len(best_fitnesses)), best_fitnesses, label=\"Best\")\n",
    "            plt.legend(loc=\"upper left\")\n",
    "            plt.show()\n",
    "\n",
    "        else:\n",
    "            plt.subplots(1, 1, figsize=(8, 8))\n",
    "            plt.title(\"Fitness by Generation\")\n",
    "            plt.plot(range(len(avg_fitnesses)), avg_fitnesses, label=\"Average\")\n",
    "            plt.plot(range(len(best_fitnesses)), best_fitnesses, label=\"Best\")\n",
    "            plt.legend(loc=\"upper left\")\n",
    "            plt.show()\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def build_tree(gene: str) -> Node:\n",
    "        \"\"\"\n",
    "        Constructs an expression tree from a gene.\n",
    "\n",
    "        :param gene: gene to turn into expression tree\n",
    "        :return: anytree Node of the root of the tree\n",
    "        \"\"\"\n",
    "\n",
    "        # shortcut to get the number of arguments to a function\n",
    "        def args(f: str) -> int:\n",
    "            return Chromosome.functions[f][\"args\"] if f in Chromosome.functions else 0\n",
    "\n",
    "        # recursively build chromosome tree\n",
    "        def grab_children(parent: Node, current_level = 1):\n",
    "            nonlocal levels\n",
    "            if current_level < len(levels):\n",
    "                nargs = args(parent.name)\n",
    "                for i in range(nargs):\n",
    "                    current_node = Node(levels[current_level][i], parent=parent)\n",
    "                    grab_children(parent=current_node, current_level=current_level + 1)\n",
    "                    if current_level < len(levels) - 1:\n",
    "                        levels[current_level + 1] = levels[current_level + 1][args(current_node.name):]\n",
    "\n",
    "        # build each level of the tree\n",
    "        levels = [gene[0]]\n",
    "        index = 0\n",
    "        while index < len(gene) and sum([args(f) for f in levels[-1]]) != 0:\n",
    "            nargs = sum([args(f) for f in levels[-1]])\n",
    "            levels.append(gene[index + 1: index + 1 + nargs])\n",
    "            index += nargs\n",
    "\n",
    "        # intialize tree and parse\n",
    "        tree = Node(gene[0])\n",
    "        grab_children(tree)\n",
    "        return tree\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    # TODO - verify recursive linking with non-commutative linking functions (e.g. -)\n",
    "    def link(*args) -> Node:\n",
    "        \"\"\"\n",
    "        Links two trees at their roots using the specified linking function.\n",
    "        Linking function must take as many arguments as number of args provided.\n",
    "\n",
    "        :param args: expression trees to link. Must be at least as many expression trees as linking function has arguments.\n",
    "        :return: expression tree with tree1 and tree2 as subtrees\n",
    "        \"\"\"\n",
    "\n",
    "        if Chromosome.linking_function not in Chromosome.functions:\n",
    "            raise ValueError(\"Linking function is not defined in Chromosome.functions.\")\n",
    "        if not all([isinstance(arg, Node) for arg in args]):\n",
    "            raise TypeError(\"Can only link expression trees.\")\n",
    "\n",
    "        nargs = Chromosome.functions[Chromosome.linking_function][\"args\"]\n",
    "\n",
    "        def link_recursive(*args) -> Node:\n",
    "            root = Node(Chromosome.linking_function)\n",
    "            if len(args) == nargs:\n",
    "                for tree in args:\n",
    "                    tree.parent = root\n",
    "                return root\n",
    "            else:\n",
    "                return link_recursive(link_recursive(*args[:nargs]), *args[nargs:])\n",
    "\n",
    "        return link_recursive(*args)\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    # TODO - calculate using numpy arrays for speed\n",
    "    def absolute_fitness(M: float, *args) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        Calculate absolute fitness of an arbitrary number of Chromosomes.\n",
    "\n",
    "        :param M: range of fitness function over domain\n",
    "        :param args: any number of gene objects\n",
    "        :return: list of fitnesses of corresponding chromosomes\n",
    "        \"\"\"\n",
    "        fitnesses = []\n",
    "        for chromosome in args:\n",
    "            # memoize fitness values\n",
    "            if chromosome._fitness_ is not None:\n",
    "                fitnesses.append(chromosome._fitness_)\n",
    "            else:\n",
    "                fitness = 0\n",
    "                for j in range(len(Chromosome.fitness_cases)):\n",
    "                    C_ij = chromosome.evaluate(Chromosome.fitness_cases[j][0])\n",
    "                    # assign any chromosome that divides by zero a fitness value of zero\n",
    "                    if type(C_ij) == np.complex or np.isnan(C_ij) or np.isinf(C_ij) or np.isneginf(C_ij):\n",
    "                        fitness = 0\n",
    "                        break\n",
    "                    T_j = Chromosome.fitness_cases[j][1]\n",
    "                    fitness += M - abs(C_ij - T_j)\n",
    "                chromosome._fitness_ = fitness\n",
    "                fitnesses.append(fitness)\n",
    "        return np.asarray(fitnesses)\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    # TODO - calculate using numpy arrays for speed\n",
    "    def relative_fitness(M: float, *args) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        Calculate relative fitness of an arbitrary number of genes.\n",
    "\n",
    "        :param M: range of fitness function over domain\n",
    "        :param args: any number of gene objects\n",
    "        :return: list of fitnesses of corresponding genes\n",
    "        \"\"\"\n",
    "        fitnesses = []\n",
    "        for chromosome in args:\n",
    "            # memoize fitness values\n",
    "            if chromosome._fitness_ is not None:\n",
    "                fitnesses.append(chromosome._fitness_)\n",
    "            else:\n",
    "                fitness = 0\n",
    "                for j in range(len(Chromosome.fitness_cases)):\n",
    "                    C_ij = chromosome.evaluate(Chromosome.fitness_cases[j][0])\n",
    "                    T_j = Chromosome.fitness_cases[j][1]\n",
    "                    fitness += M - 100*abs(C_ij / T_j - 1)\n",
    "                chromosome._fitness_ = fitness\n",
    "                fitnesses.append(fitness)\n",
    "        return np.asarray(fitnesses)\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def inv_squared_error(*args) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        Classical 1/(1+(squared error) fitness value.\n",
    "\n",
    "        :param args: list of chromosomes to calculate fitness of\n",
    "        :return: ndarray of fitness values for each given chromosome\n",
    "        \"\"\"\n",
    "        fitnesses = []\n",
    "        for chromosome in args:\n",
    "            # memoize fitness values\n",
    "            if chromosome._fitness_ is not None:\n",
    "                fitnesses.append(chromosome._fitness_)\n",
    "            else:\n",
    "                fitness = 0\n",
    "                for j in range(len(Chromosome.fitness_cases)):\n",
    "                    C_ij = chromosome.evaluate(Chromosome.fitness_cases[j][0])\n",
    "                    if type(C_ij) == np.complex or np.isnan(C_ij) or np.isinf(C_ij) or np.isneginf(C_ij):\n",
    "                        fitness = np.inf\n",
    "                        break\n",
    "                    T_j = Chromosome.fitness_cases[j][1]\n",
    "                    fitness += (C_ij - T_j)**2\n",
    "                chromosome._fitness_ = 1.0/(1+fitness)\n",
    "                fitnesses.append(chromosome._fitness_)\n",
    "        return np.asarray(fitnesses)\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def centralized_inv_squared_error(center: float, dimension: str, *args) -> np.ndarray:\n",
    "        \"\"\"\n",
    "        Mostly unused, a fitness function that focuses on error near a given point.\n",
    "\n",
    "        :param center: point around which the fitness values are more important\n",
    "        :param dimension: which dimension of the fitness case to use, in the case of multivariate functions\n",
    "        :param args: any chromosomes to calculate fitness of\n",
    "        :return: ndarray of fitness values\n",
    "        \"\"\"\n",
    "        fitnesses = []\n",
    "        for chromosome in args:\n",
    "            # memoize fitness values\n",
    "            if chromosome._fitness_ is not None:\n",
    "                fitnesses.append(chromosome._fitness_)\n",
    "            else:\n",
    "                fitness = 0\n",
    "                for j in range(len(Chromosome.fitness_cases)):\n",
    "                    C_ij = chromosome.evaluate(Chromosome.fitness_cases[j][0])\n",
    "                    if type(C_ij) == np.complex or np.isnan(C_ij) or np.isinf(C_ij) or np.isneginf(C_ij):\n",
    "                        fitness = np.inf\n",
    "                        break\n",
    "                    T_j = Chromosome.fitness_cases[j][1]\n",
    "                    fitness += abs(C_ij - T_j)**(1/abs(Chromosome.fitness_cases[j][0][dimension] - center))\n",
    "                chromosome._fitness_ = 1.0 / (1 + fitness)\n",
    "                fitnesses.append(chromosome._fitness_)\n",
    "        return np.asarray(fitnesses)\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def generate_random_gene() -> str:\n",
    "        \"\"\"\n",
    "        Generates one random gene based on settings specified in Chromosome class.\n",
    "        :return: string of valid characters\n",
    "        \"\"\"\n",
    "        possible_chars = list(Chromosome.functions.keys()) + Chromosome.terminals\n",
    "        head = \"\".join([possible_chars[randint(0, len(possible_chars) - 1)] for _ in range(Chromosome.head_length)])\n",
    "        tail = \"\".join([Chromosome.terminals[randint(0, len(Chromosome.terminals) - 1)] for _ in range(Chromosome.length - Chromosome.head_length)])\n",
    "        return head + tail\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def generate_random_individual() -> 'Chromosome':\n",
    "        \"\"\"\n",
    "        Generates one random individual based on settings specified in Chromosome class.\n",
    "        :return: new Chromosome\n",
    "        \"\"\"\n",
    "        return Chromosome([Chromosome.generate_random_gene() for _ in range(Chromosome.num_genes)])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GeneExpressionProgram Class - Container for Program Execution\n",
    "Why do you have a class where every method is static?\n",
    "\n",
    "Because I don't like polluting the global namespace, and I like grouping variables by their function.\n",
    "\n",
    "Main driver for Gene Expression Programming. Contains methods for:\n",
    "* Evolving a configured program\n",
    "* Roulette wheel selection\n",
    "* Mutation\n",
    "* All kinds of recombination\n",
    "* Plotting the results of evolution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "class GeneExpressionProgram:\n",
    "\n",
    "    ### Hyperparameters ###\n",
    "    NUM_RUNS = 5\n",
    "    NUM_GENERATIONS = 500\n",
    "    POPULATION_SIZE = 100\n",
    "    NUM_FITNESS_CASES = 10\n",
    "    ERROR_TOLERANCE = 0.0000001\n",
    "\n",
    "    ### Reproduction ###\n",
    "    MUTATION_RATE = 0.051\n",
    "    ONE_POINT_CROSSOVER_RATE, TWO_POINT_CROSSOVER_RATE, GENE_CROSSOVER_RATE = 0.2, 0.5, 0.1\n",
    "    IS_TRANSPOSITION_RATE, IS_ELEMENTS_LENGTH = 0.1, [1, 2, 3]\n",
    "    RIS_TRANSPOSITION_RATE, RIS_ELEMENTS_LENGTH = 0.1, [1, 2, 3]\n",
    "    GENE_TRANSPOSITION_RATE = 0.1\n",
    "\n",
    "    ### Fitness Evaluation ###\n",
    "    OBJECTIVE_FUNCTION = None\n",
    "    FITNESS_FUNCTION = None\n",
    "    FITNESS_FUNCTION_ARGS = list()\n",
    "    OBJECTIVE_MIN, OBJECTIVE_MAX = None, None\n",
    "    FUNCTION_Y_RANGE = None\n",
    "\n",
    "\n",
    "    def __init__(self):\n",
    "        pass\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def evolve() -> (Chromosome, list, list):\n",
    "        \"\"\"\n",
    "        Execute Gene Expression Programming algorithm\n",
    "\n",
    "        :return: tuple of:\n",
    "            the best fit chromosome,\n",
    "            list of average fitness by generation (for plotting),\n",
    "            list of best fitness by generation (for plotting\n",
    "        \"\"\"\n",
    "\n",
    "        # create initial population\n",
    "        population = [Chromosome.generate_random_individual() for _ in range(GeneExpressionProgram.POPULATION_SIZE)]\n",
    "\n",
    "        generation = 0\n",
    "        best_fit_individual = None\n",
    "        average_fitness_by_generation = []\n",
    "        best_fitness_by_generation = []\n",
    "        while generation < GeneExpressionProgram.NUM_GENERATIONS:\n",
    "\n",
    "            ### EVALUATION ###\n",
    "\n",
    "            # calcluate fitnesses for population\n",
    "            population_fitnesses = GeneExpressionProgram.FITNESS_FUNCTION(*GeneExpressionProgram.FITNESS_FUNCTION_ARGS, *population)\n",
    "\n",
    "            # find best fit individual\n",
    "            # noinspection PyTypeChecker\n",
    "            best_fit_generation = population[np.argmax(population_fitnesses)]\n",
    "            if generation == 0 or best_fit_individual.fitness() < best_fit_generation.fitness():\n",
    "                best_fit_individual = deepcopy(best_fit_generation)\n",
    "\n",
    "            # skip rest of loop if we have found optimal solution\n",
    "            if abs(best_fit_individual.fitness() - Chromosome.max_fitness) <= GeneExpressionProgram.ERROR_TOLERANCE:\n",
    "                average_fitness_generation = float(np.mean(population_fitnesses))\n",
    "                average_fitness_by_generation.append(average_fitness_generation)\n",
    "                best_fitness_by_generation.append(best_fit_individual.fitness())\n",
    "                break\n",
    "\n",
    "            next_generation = list()\n",
    "\n",
    "\n",
    "            ### SELECTION (roulette wheel with simple elitism) ###\n",
    "\n",
    "            # copy best individual to next generation\n",
    "            next_generation.append(deepcopy(best_fit_individual))\n",
    "\n",
    "            # select the rest of the next generation with roulette wheel selection\n",
    "            all_parents = GeneExpressionProgram.roulette_wheel_selection(population, len(population))\n",
    "\n",
    "            # Mutation\n",
    "            all_parents = list(map(GeneExpressionProgram.mutate, all_parents))\n",
    "\n",
    "            # IS Transposition\n",
    "            all_parents = list(map(GeneExpressionProgram.is_transposition, all_parents))\n",
    "\n",
    "            # RIS Transposition\n",
    "            all_parents = list(map(GeneExpressionProgram.ris_transposition, all_parents))\n",
    "\n",
    "            # Gene Transposition\n",
    "            all_parents = list(map(GeneExpressionProgram.gene_transposition, all_parents))\n",
    "\n",
    "            # Recombination\n",
    "            shuffle(all_parents)\n",
    "            for i in range(1, GeneExpressionProgram.POPULATION_SIZE, 2):\n",
    "\n",
    "                # in case we don't have a pair to check for crossover, avoid index error\n",
    "                if i + 1 >= GeneExpressionProgram.POPULATION_SIZE:\n",
    "                    next_generation.append(all_parents[i])\n",
    "                    break\n",
    "\n",
    "                child1, child2 = all_parents[i], all_parents[i+1]\n",
    "\n",
    "                # One-point Recombination\n",
    "                if random() < GeneExpressionProgram.ONE_POINT_CROSSOVER_RATE:\n",
    "                    child1, child2 = GeneExpressionProgram.one_point_recombination(child1, child2)\n",
    "\n",
    "                # Two-point Recombination\n",
    "                elif random() < GeneExpressionProgram.TWO_POINT_CROSSOVER_RATE:\n",
    "                    child1, child2 = GeneExpressionProgram.two_point_recombination(child1, child2)\n",
    "\n",
    "                # Gene Recombination\n",
    "                elif random() < GeneExpressionProgram.GENE_CROSSOVER_RATE:\n",
    "                    child1, child2 = GeneExpressionProgram.gene_recombination(child1, child2)\n",
    "\n",
    "                # Include children in next generation\n",
    "                next_generation.append(child1)\n",
    "                next_generation.append(child2)\n",
    "\n",
    "            # prepare for next iteration\n",
    "            population = next_generation\n",
    "            generation += 1\n",
    "\n",
    "            average_fitness_generation = float(np.mean(population_fitnesses))\n",
    "            average_fitness_by_generation.append(average_fitness_generation)\n",
    "            best_fitness_by_generation.append(best_fit_individual.fitness())\n",
    "            print(\"Generation: %d\\tPopulation Size: %d\\tAverage Fitness: %.5f\\tBest Fitness (overall): %.5f\" %\n",
    "                  (generation, len(population), average_fitness_generation, best_fit_individual.fitness()))\n",
    "\n",
    "        return best_fit_individual, average_fitness_by_generation, best_fitness_by_generation\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def random_search(num_generations: int, fitness_function: callable, fitness_function_args: list) -> (Chromosome, list, list):\n",
    "        best = None\n",
    "        best_fitness = 0\n",
    "        average_fitnesses = []\n",
    "        best_fitnesses = []\n",
    "        for gen in range(num_generations):\n",
    "            generation_fitnesses = []\n",
    "            for individual in range(GeneExpressionProgram.POPULATION_SIZE):\n",
    "                current = Chromosome.generate_random_individual()\n",
    "                current_fitness = fitness_function(*fitness_function_args, current)\n",
    "                generation_fitnesses.append(current_fitness)\n",
    "                if best is None or best_fitness <= current_fitness:\n",
    "                    best = deepcopy(current)\n",
    "                    best_fitness = best.fitness()\n",
    "            average_fitnesses.append(np.mean(generation_fitnesses))\n",
    "            best_fitnesses.append(best_fitness)\n",
    "            if best_fitness >= Chromosome.max_fitness:\n",
    "                break\n",
    "        return best, average_fitnesses, best_fitnesses\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def roulette_wheel_selection(chromosomes: list, n: int) -> list:\n",
    "        \"\"\"\n",
    "        Returns n randomly selected chromosomes using roulette wheel selection.\n",
    "        Adapted from:\n",
    "        https://stackoverflow.com/questions/2140787/select-k-random-elements-from-a-list-whose-elements-have-weights\n",
    "\n",
    "        :param chromosomes: list of chromosomes to select from\n",
    "        :param n: number of samples to retrieve\n",
    "        :return: list of chosen chromosome(s)\n",
    "        \"\"\"\n",
    "        total = float(sum(c.fitness() for c in chromosomes))\n",
    "        i = 0\n",
    "        fitness = chromosomes[0].fitness()\n",
    "        while n:\n",
    "            x = total * (1 - random() ** (1.0 / n))\n",
    "            total -= x\n",
    "            while x > fitness:\n",
    "                x -= fitness\n",
    "                i += 1\n",
    "                fitness = chromosomes[i].fitness()\n",
    "            fitness -= x\n",
    "            yield chromosomes[i]\n",
    "            n -= 1\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def mutate(chromosome: Chromosome) -> Chromosome:\n",
    "        \"\"\"\n",
    "        Randomly mutate genes in a chromosome.\n",
    "\n",
    "        :param chromosome: chromosome to mutate\n",
    "        :return: mutated chromosome\n",
    "        \"\"\"\n",
    "\n",
    "        head_characters = list(Chromosome.functions.keys()) + Chromosome.terminals\n",
    "\n",
    "        new_genes = []\n",
    "        # for each gene (multigenic chromosomes)\n",
    "        for gene in range(Chromosome.num_genes):\n",
    "            new_gene = \"\"\n",
    "            # for each character in the gene\n",
    "            for i in range(len(chromosome.genes[gene])):\n",
    "                # if we are to mutate it\n",
    "                if random() < GeneExpressionProgram.MUTATION_RATE:\n",
    "                    # if we are mutating the head\n",
    "                    if i < Chromosome.head_length:\n",
    "                        new_gene += head_characters[randint(0, len(head_characters) - 1)]\n",
    "                    else:\n",
    "                        new_gene += Chromosome.terminals[randint(0, len(Chromosome.terminals) - 1)]\n",
    "                else:\n",
    "                    new_gene += chromosome.genes[gene][i]\n",
    "            new_genes.append(new_gene)\n",
    "\n",
    "        # create new chromosome to ensure memoized fitness values are recalculated.\n",
    "        new_chromosome = Chromosome(new_genes)\n",
    "        new_chromosome.ephemeral_random_constants = chromosome.ephemeral_random_constants\n",
    "\n",
    "        # mutate ephemeral random constants\n",
    "        for constant in range(len(new_chromosome.ephemeral_random_constants)):\n",
    "            if random() < GeneExpressionProgram.MUTATION_RATE:\n",
    "                new_chromosome.ephemeral_random_constants[constant] = np.random.uniform(*Chromosome.ephemeral_random_constants_range)\n",
    "\n",
    "        return new_chromosome\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def is_transposition(chromosome: Chromosome) -> Chromosome:\n",
    "        \"\"\"\n",
    "        Insertion Sequence transposition.\n",
    "\n",
    "        :param chromosome: chromosome to perform IS transposition on\n",
    "        :return: new chromosome\n",
    "        \"\"\"\n",
    "\n",
    "        # TODO - properly transpose ephemeral random constants\n",
    "\n",
    "        if random() < GeneExpressionProgram.IS_TRANSPOSITION_RATE:\n",
    "\n",
    "            # determine parameters of transposition\n",
    "            length = np.random.choice(GeneExpressionProgram.IS_ELEMENTS_LENGTH)\n",
    "            source_gene = randint(0, len(chromosome.genes) - 1)\n",
    "            target_gene = randint(0, len(chromosome.genes) - 1)\n",
    "            target_position = randint(1, Chromosome.head_length - length)\n",
    "            sequence_start = randint(0, len(chromosome.genes[source_gene]))\n",
    "\n",
    "            transposition_string = chromosome.genes[source_gene][sequence_start:min(Chromosome.length, sequence_start + length)]\n",
    "\n",
    "            # make substitution\n",
    "            new_chromosome = Chromosome(chromosome.genes)\n",
    "            new_chromosome.ephemeral_random_constants = chromosome.ephemeral_random_constants\n",
    "            new_chromosome.genes[target_gene] = new_chromosome.genes[target_gene][:target_position] + \\\n",
    "                                                transposition_string + \\\n",
    "                                                new_chromosome.genes[target_gene][target_position + length:]\n",
    "\n",
    "            return new_chromosome\n",
    "\n",
    "        else:\n",
    "\n",
    "            return chromosome\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def ris_transposition(chromosome: Chromosome) -> Chromosome:\n",
    "        \"\"\"\n",
    "        Root Insertion Sequence transposition.\n",
    "\n",
    "        :param chromosome: chromosome to perform RIS transposition on\n",
    "        :return: new chromosome\n",
    "        \"\"\"\n",
    "\n",
    "        # TODO - properly transpose ephemeral random constants\n",
    "\n",
    "        start_point = randint(0, Chromosome.head_length - 1)\n",
    "        gene = randint(0, Chromosome.num_genes - 1)\n",
    "        while start_point < Chromosome.head_length and chromosome.genes[gene][start_point] not in Chromosome.functions:\n",
    "            start_point += 1\n",
    "\n",
    "        if random() < GeneExpressionProgram.RIS_TRANSPOSITION_RATE and chromosome.genes[gene][start_point] in Chromosome.functions:\n",
    "            ris_length = np.random.choice(GeneExpressionProgram.RIS_ELEMENTS_LENGTH)\n",
    "            ris_string = chromosome.genes[gene][start_point:start_point+ris_length]\n",
    "\n",
    "            new_chromosome = Chromosome(chromosome.genes)\n",
    "            new_chromosome.ephemeral_random_constants = chromosome.ephemeral_random_constants\n",
    "            old_head = new_chromosome.genes[gene][:Chromosome.head_length]\n",
    "            new_head = old_head[:start_point] + ris_string + old_head[start_point:]\n",
    "            new_chromosome.genes[gene] = new_head[:Chromosome.head_length] + new_chromosome.genes[gene][Chromosome.head_length:]\n",
    "\n",
    "            return new_chromosome\n",
    "\n",
    "        else:\n",
    "\n",
    "            return chromosome\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def gene_transposition(chromosome: Chromosome) -> Chromosome:\n",
    "        \"\"\"\n",
    "        Gene Insertion Sequence transposition.\n",
    "\n",
    "        :param chromosome: chromosome to perform gene transposition on\n",
    "        :return: new chromosome\n",
    "        \"\"\"\n",
    "\n",
    "        # TODO - properly transpose ephemeral random constants\n",
    "\n",
    "        if Chromosome.num_genes > 1 and random() < GeneExpressionProgram.GENE_TRANSPOSITION_RATE:\n",
    "\n",
    "            index = randint(0, Chromosome.num_genes - 1)\n",
    "            temp = chromosome.genes[index]\n",
    "            chromosome.genes[index] = chromosome.genes[0]\n",
    "            chromosome.genes[0] = temp\n",
    "            new_chromosome = Chromosome(chromosome.genes)\n",
    "            new_chromosome.ephemeral_random_constants = chromosome.ephemeral_random_constants\n",
    "            return new_chromosome\n",
    "\n",
    "        else:\n",
    "\n",
    "            return chromosome\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def one_point_recombination(chromosome1: Chromosome, chromosome2: Chromosome) -> (Chromosome, Chromosome):\n",
    "        \"\"\"\n",
    "        Classical one point recombination.\n",
    "\n",
    "        :param chromosome1: parent 1\n",
    "        :param chromosome2: parent 2\n",
    "        :return: offspring 1, offspring 2\n",
    "        \"\"\"\n",
    "        gene = randint(0, Chromosome.num_genes - 1)\n",
    "        position = randint(0, Chromosome.length)\n",
    "\n",
    "        child1_split_gene = chromosome1.genes[gene][:position] + chromosome2.genes[gene][position:]\n",
    "        child2_split_gene = chromosome2.genes[gene][:position] + chromosome1.genes[gene][position:]\n",
    "\n",
    "        child1_genes = chromosome1.genes[:gene] + [child1_split_gene] + (chromosome2.genes[gene+1:] if gene < Chromosome.num_genes - 1 else [])\n",
    "        child2_genes = chromosome2.genes[:gene] + [child2_split_gene] + (chromosome1.genes[gene + 1:] if gene < Chromosome.num_genes - 1 else [])\n",
    "\n",
    "        child1, child2 = Chromosome(child1_genes), Chromosome(child2_genes)\n",
    "\n",
    "        constants_split_position = randint(0, Chromosome.length - 1)\n",
    "        child1.ephemeral_random_constants = chromosome1.ephemeral_random_constants[:constants_split_position] + \\\n",
    "                                            chromosome2.ephemeral_random_constants[constants_split_position:]\n",
    "\n",
    "        child2.ephemeral_random_constants = chromosome2.ephemeral_random_constants[:constants_split_position] + \\\n",
    "                                            chromosome1.ephemeral_random_constants[constants_split_position:]\n",
    "\n",
    "        return child1, child2\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def two_point_recombination(chromosome1: Chromosome, chromosome2: Chromosome) -> (Chromosome, Chromosome):\n",
    "        \"\"\"\n",
    "        Classical two point recombination.\n",
    "\n",
    "        :param chromosome1: parent 1\n",
    "        :param chromosome2: parent 2\n",
    "        :return: offspring 1, offsprint 2\n",
    "        \"\"\"\n",
    "\n",
    "        # generate crossover points\n",
    "        position1, position2 = sorted([randint(0, Chromosome.length*Chromosome.num_genes - 1), randint(0, Chromosome.length*Chromosome.num_genes - 1)])\n",
    "\n",
    "        # join genes into single string for ease of manipulation\n",
    "        child1_genes_str = \"\".join(chromosome1.genes)\n",
    "        child2_genes_str = \"\".join(chromosome2.genes)\n",
    "\n",
    "        # perform crossover\n",
    "        child1_genes = child1_genes_str[:position1] + child2_genes_str[position1:position2] + child1_genes_str[position2:]\n",
    "        child2_genes = child2_genes_str[:position1] + child1_genes_str[position1:position2] + child2_genes_str[position2:]\n",
    "\n",
    "        # split genes from string into list\n",
    "        child1_genes = [child1_genes[i:i + Chromosome.length] for i in range(0, Chromosome.num_genes * Chromosome.length, Chromosome.length)]\n",
    "        child2_genes = [child2_genes[i:i + Chromosome.length] for i in range(0, Chromosome.num_genes * Chromosome.length, Chromosome.length)]\n",
    "\n",
    "        child1, child2 = Chromosome(child1_genes), Chromosome(child2_genes)\n",
    "        split_positions = sorted([randint(0, Chromosome.length - 1), randint(0, Chromosome.length - 1)])\n",
    "\n",
    "        child1.ephemeral_random_constants = chromosome1.ephemeral_random_constants[:split_positions[0]] + \\\n",
    "                                            chromosome2.ephemeral_random_constants[split_positions[0]:split_positions[1]] + \\\n",
    "                                            chromosome1.ephemeral_random_constants[split_positions[1]:]\n",
    "        child2.ephemeral_random_constants = chromosome2.ephemeral_random_constants[:split_positions[0]] + \\\n",
    "                                            chromosome1.ephemeral_random_constants[split_positions[0]:split_positions[1]] + \\\n",
    "                                            chromosome2.ephemeral_random_constants[split_positions[1]:]\n",
    "        return child1, child2\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def gene_recombination(chromosome1: Chromosome, chromosome2: Chromosome) -> (Chromosome, Chromosome):\n",
    "        \"\"\"\n",
    "        Two point recombination that occurs along gene boundaries for multigenic chromosomes.\n",
    "\n",
    "        :param chromosome1: parent 1\n",
    "        :param chromosome2: parent 2\n",
    "        :return: offspring 1, offspring 2\n",
    "        \"\"\"\n",
    "\n",
    "        # choose gene to swap\n",
    "        gene = randint(0, Chromosome.num_genes - 1)\n",
    "\n",
    "        # initialize children genes\n",
    "        child1_genes = chromosome1.genes\n",
    "        child2_genes = chromosome2.genes\n",
    "\n",
    "        # perform swap\n",
    "        child1_genes[gene] = chromosome2.genes[gene]\n",
    "        child2_genes[gene] = chromosome1.genes[gene]\n",
    "\n",
    "        return Chromosome(child1_genes), Chromosome(child2_genes)\n",
    "\n",
    "\n",
    "    @staticmethod\n",
    "    def plot_reps(avg_fitnesses: list, best_fitnesses: list, random_search_avg: list = None, random_search_best: list = None) -> None:\n",
    "        \"\"\"\n",
    "        Plot all reps with global best solutions.\n",
    "\n",
    "        :param avg_fitnesses: list of lists containing average fitnesses by generation for each rep\n",
    "        :param best_fitnesses: same as avg_fitnesses but best fitnesses\n",
    "        :param random_search_avg: average fitness value for random search across generations\n",
    "        :param random_search_best: best fitness value for random search by generation\n",
    "        :return: void\n",
    "        \"\"\"\n",
    "\n",
    "        is_random_search = not (random_search_avg is None or random_search_best is None)\n",
    "\n",
    "        plt.subplots(1, 2, figsize=(16, 8))\n",
    "\n",
    "        plt.subplot(1, 2, 1)\n",
    "        plt.title(\"Average Fitness by Generation\")\n",
    "        plt.xlabel(\"Generation\")\n",
    "        plt.ylabel(\"Average Fitness\")\n",
    "\n",
    "        # plot each rep\n",
    "        for rep in range(GeneExpressionProgram.NUM_RUNS):\n",
    "            plt.plot(range(len(avg_fitnesses[rep])), avg_fitnesses[rep], label=\"Rep %d Average\" % (rep + 1))\n",
    "\n",
    "        if is_random_search:\n",
    "            plt.plot(range(len(random_search_avg)), random_search_avg, label=\"Random Search Average\")\n",
    "            #plt.plot(range(len(random_search_best)), random_search_best, label=\"Random Search Best\")\n",
    "\n",
    "        plt.legend(loc=\"upper left\")\n",
    "\n",
    "        plt.subplot(1, 2, 2)\n",
    "        plt.title(\"Best Fitness by Generation\")\n",
    "        plt.xlabel(\"Generation\")\n",
    "        plt.ylabel(\"Best Fitness\")\n",
    "\n",
    "        # plot each rep\n",
    "        for rep in range(GeneExpressionProgram.NUM_RUNS):\n",
    "            plt.plot(range(len(best_fitnesses[rep])), best_fitnesses[rep], label=\"Rep %d Best\" % (rep + 1))\n",
    "\n",
    "        if is_random_search:\n",
    "            plt.plot(range(len(random_search_best)), random_search_best, label=\"Random Search Best\")\n",
    "\n",
    "        plt.legend(loc=\"upper left\")\n",
    "        plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Symbolic Regression: $f(a) = a^4 + a^3 + a^2 + a$\n",
    "Run this function to have GEP attempt to discover the function\n",
    "$$ f(a) = a^4 + a^3 + a^2 + a $$\n",
    "across 10 fitness cases. The fitness cases are of the form $\\left(a, f(a)\\right)$ for randomly generated $a$ between `OBJECTIVE_MIN` and `OBJECTIVE_MAX`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def a4_a3_a2_a1():\n",
    "\n",
    "    # define objective function\n",
    "    GeneExpressionProgram.OBJECTIVE_FUNCTION = staticmethod(lambda a: a ** 4 + a ** 3 + a ** 2 + a)\n",
    "    GeneExpressionProgram.OBJECTIVE_MIN, GeneExpressionProgram.OBJECTIVE_MAX = 0, 20\n",
    "\n",
    "    # Define terminals and functions\n",
    "    Chromosome.terminals = [\"a\"]\n",
    "    Chromosome.functions = {\n",
    "        \"+\": {\"args\": 2, \"f\": lambda x, y: x + y},\n",
    "        \"-\": {\"args\": 2, \"f\": lambda x, y: x - y},\n",
    "        \"*\": {\"args\": 2, \"f\": lambda x, y: x * y},\n",
    "        \"/\": {\"args\": 2, \"f\": lambda x, y: x / y}\n",
    "    }\n",
    "    Chromosome.fitness_cases = [\n",
    "        ({\"a\": a}, GeneExpressionProgram.OBJECTIVE_FUNCTION(a))\n",
    "        for a in np.random.rand(GeneExpressionProgram.NUM_FITNESS_CASES) * (\n",
    "        GeneExpressionProgram.OBJECTIVE_MAX - GeneExpressionProgram.OBJECTIVE_MIN) + GeneExpressionProgram.OBJECTIVE_MIN\n",
    "    ]\n",
    "\n",
    "    GeneExpressionProgram.FUNCTION_Y_RANGE = \\\n",
    "        GeneExpressionProgram.OBJECTIVE_FUNCTION(GeneExpressionProgram.OBJECTIVE_MAX) - \\\n",
    "        GeneExpressionProgram.OBJECTIVE_FUNCTION(GeneExpressionProgram.OBJECTIVE_MIN)\n",
    "\n",
    "    Chromosome.linking_function = \"+\"\n",
    "    Chromosome.max_fitness = 1 #GeneExpressionProgram.FUNCTION_Y_RANGE * GeneExpressionProgram.NUM_FITNESS_CASES\n",
    "\n",
    "    ans = GeneExpressionProgram.evolve()\n",
    "    ans.print_tree()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Symbolic Regression: $f(x) = \\sin(x)$\n",
    "Watch as GEP attempts to approximate $f(x) = sin(x)$ using only addition, subtraction, multiplication, division, the variable $x$, and the number 6! If you're lucky, it may come up with the second order Taylor approximation centered at $x=0$:\n",
    "$$\\sin(x) \\approx x - \\frac{x^3}{6}$$\n",
    "However, it often finds the first order Taylor approximation, calls it good enough, and quits. The first order Taylor approximation is\n",
    "$$\\sin(x) \\approx x$$\n",
    "for $x \\approx 0$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def sinx_polynomial():\n",
    "    # define objective function\n",
    "    GeneExpressionProgram.OBJECTIVE_FUNCTION = staticmethod(lambda a: np.sin(a))\n",
    "    GeneExpressionProgram.OBJECTIVE_MIN, GeneExpressionProgram.OBJECTIVE_MAX = -5, 5\n",
    "    GeneExpressionProgram.FITNESS_FUNCTION = Chromosome.centralized_inv_squared_error\n",
    "    GeneExpressionProgram.FITNESS_FUNCTION_ARGS = [0, \"x\"]\n",
    "    GeneExpressionProgram.NUM_FITNESS_CASES = 10\n",
    "    GeneExpressionProgram.NUM_GENERATIONS = 150\n",
    "    GeneExpressionProgram.POPULATION_SIZE = 200\n",
    "    GeneExpressionProgram.MUTATION_RATE = 0.05\n",
    "\n",
    "    Chromosome.head_length = 8\n",
    "    Chromosome.num_genes = 3\n",
    "    Chromosome.length = 24\n",
    "\n",
    "    # Define terminals and functions\n",
    "    Chromosome.terminals = [\"x\", \"6\", \"!\"]\n",
    "    Chromosome.constants = {\"!\": 120}\n",
    "    Chromosome.functions = {\n",
    "        \"+\": {\"args\": 2, \"f\": lambda x, y: x + y},\n",
    "        \"-\": {\"args\": 2, \"f\": lambda x, y: x - y},\n",
    "        \"*\": {\"args\": 2, \"f\": lambda x, y: x * y},\n",
    "        \"/\": {\"args\": 2, \"f\": lambda x, y: x / y}\n",
    "        #\"^\": {\"args\": 2, \"f\": lambda x, y: x ** y}\n",
    "    }\n",
    "    Chromosome.fitness_cases = [\n",
    "        ({\"x\": x}, GeneExpressionProgram.OBJECTIVE_FUNCTION(x))\n",
    "        for x in np.random.rand(GeneExpressionProgram.NUM_FITNESS_CASES) * (\n",
    "        GeneExpressionProgram.OBJECTIVE_MAX - GeneExpressionProgram.OBJECTIVE_MIN) + GeneExpressionProgram.OBJECTIVE_MIN\n",
    "    ]\n",
    "\n",
    "    GeneExpressionProgram.FUNCTION_Y_RANGE = 2\n",
    "    Chromosome.linking_function = \"*\"\n",
    "    Chromosome.max_fitness = 1 #GeneExpressionProgram.FUNCTION_Y_RANGE * GeneExpressionProgram.NUM_FITNESS_CASES\n",
    "\n",
    "    ans, average_fitnesses, best_fitnesses = GeneExpressionProgram.evolve()\n",
    "    ans.print_tree()\n",
    "    ans.plot_solution(GeneExpressionProgram.OBJECTIVE_FUNCTION,\n",
    "                      GeneExpressionProgram.OBJECTIVE_MIN, GeneExpressionProgram.OBJECTIVE_MAX,\n",
    "                      average_fitnesses, best_fitnesses)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CartPole v1\n",
    "GEP evolves a real-valued function which is positive when the cart should move to the right and negative when the cart should move to the left."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def cart_pole_real():\n",
    "\n",
    "    import gym\n",
    "    env = gym.make('CartPole-v1')\n",
    "\n",
    "    def fitness(num_trials: int, render: bool = False, doPrint = False, *args) -> np.ndarray:\n",
    "        fitnesses = []\n",
    "        for chromosome_index in range(len(args)):\n",
    "            chromosome = args[chromosome_index]\n",
    "            total_reward = 0\n",
    "            if doPrint:\n",
    "                print(\"\\tCalculating chromosome %d across %d trials.\" % (chromosome_index, num_trials), end=\"\\t\")\n",
    "            for trial in range(num_trials):\n",
    "                x, theta, dx, dtheta = env.reset()\n",
    "                t = 0\n",
    "                while True:\n",
    "                    if render: env.render()\n",
    "                    action = chromosome.evaluate({\"x\": x, \"v\": dx, \"t\": theta, \"u\": dtheta})\n",
    "                    observation, reward, done, info = env.step(action > 0)\n",
    "                    x, theta, dx, dtheta = observation\n",
    "                    total_reward += reward\n",
    "                    if done:\n",
    "                        break\n",
    "                    t += 1\n",
    "            chromosome._fitness_ = total_reward / float(num_trials)\n",
    "            fitnesses.append(total_reward / float(num_trials))\n",
    "            if doPrint:\n",
    "                print(\"Fitness: %.5f\" % (total_reward / float(num_trials)))\n",
    "        return np.asarray(fitnesses)\n",
    "\n",
    "    Chromosome.functions = {\n",
    "        \"+\": {\"args\": 2, \"f\": lambda x, y: x + y},\n",
    "        \"-\": {\"args\": 2, \"f\": lambda x, y: x - y},\n",
    "        \"*\": {\"args\": 2, \"f\": lambda x, y: x * y},\n",
    "        \"/\": {\"args\": 2, \"f\": lambda x, y: x / y}\n",
    "    }\n",
    "\n",
    "    Chromosome.terminals = [\n",
    "        \"x\",    # cart x-value\n",
    "        \"v\",    # cart velocity\n",
    "        \"t\",    # pole angle\n",
    "        \"u\"     # rate of change of pole angle\n",
    "    ]\n",
    "\n",
    "    GeneExpressionProgram.FITNESS_FUNCTION = fitness\n",
    "    GeneExpressionProgram.FITNESS_FUNCTION_ARGS = [10, False, False]\n",
    "    GeneExpressionProgram.POPULATION_SIZE = 10\n",
    "    GeneExpressionProgram.ERROR_TOLERANCE = 0\n",
    "    GeneExpressionProgram.NUM_RUNS = 5\n",
    "    GeneExpressionProgram.NUM_GENERATIONS = 25\n",
    "\n",
    "    Chromosome.max_fitness = 500\n",
    "    Chromosome.num_genes = 3\n",
    "    Chromosome.length = 30\n",
    "    Chromosome.head_length = 10\n",
    "    Chromosome.linking_function = \"+\"\n",
    "\n",
    "    average_fitnesses = []\n",
    "    best_fitnesses = []\n",
    "\n",
    "    for rep in range(GeneExpressionProgram.NUM_RUNS):\n",
    "        ans, gen_average_fitnesses, gen_best_fitnesses = GeneExpressionProgram.evolve()\n",
    "        average_fitnesses.append(gen_average_fitnesses)\n",
    "        best_fitnesses.append(gen_best_fitnesses)\n",
    "        ans.print_tree()\n",
    "        print(ans.genes)\n",
    "\n",
    "        if fitness(100, False, True, ans)[0] == Chromosome.max_fitness:\n",
    "            fitness(5, True, False, ans)\n",
    "        \n",
    "\n",
    "    GeneExpressionProgram.plot_reps(average_fitnesses, best_fitnesses)\n",
    "\n",
    "    # Very good genes (500 fitness over 100 attempts):\n",
    "    # ['-u-uvx*txtxxutuutvtvtvttttutvu', '+*vxu/x/-/tutxtxxxuvuuxvtuvvtu', '+ut*vv*vtvxvxvuxvtxvtvuutvtttt']\n",
    "    # ['t/v*/xtttvuxtvvttutxxtuvxuvttt', '-v*+v*+t+tuxtvvttuuvxvtuvuvxtt', 'u++tu+uvtvtxtvtvtuvxxttvxvxvv']\n",
    "    # ['-t-*+u*u-*tuutxxuutxtxttxttvxu', '+-u**+-/vxvxxxvtxvxuxvxvvxxtvx', '+xv-xvx-xtuutxxtxxxtxxuttvxxut']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Hard Symbolic Regression\n",
    "GEP attempts to find the equation\n",
    "$$ f(a) = 4.251a^2 + \\ln(a^2) + 7.243e^a $$\n",
    "using ephemeral random constants across 5000 generations and 5 reps. This often gets stuck in local optima, and hence does not always attain a large fitness value. The success of this algorithm depends greatly on the luck of the initial population and the probability of favorable mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def hard_regression():\n",
    "    # TODO - crossover and mutation of constants\n",
    "    # define objective function\n",
    "    GeneExpressionProgram.OBJECTIVE_FUNCTION = staticmethod(lambda a: 4.251*np.power(a, 2) + np.log(np.power(a, 2)) + 7.243*np.exp(a))\n",
    "    GeneExpressionProgram.OBJECTIVE_MIN, GeneExpressionProgram.OBJECTIVE_MAX = -1, 1\n",
    "    GeneExpressionProgram.NUM_FITNESS_CASES = 20\n",
    "    GeneExpressionProgram.FITNESS_FUNCTION = Chromosome.inv_squared_error\n",
    "    GeneExpressionProgram.NUM_GENERATIONS = 5000\n",
    "    GeneExpressionProgram.NUM_RUNS = 5\n",
    "    GeneExpressionProgram.MUTATION_RATE = 0.041\n",
    "\n",
    "    # Define terminals and functions\n",
    "    Chromosome.terminals = [\"a\", \"?\"]\n",
    "    Chromosome.ephemeral_random_constants_range = (-1, 1)\n",
    "    Chromosome.functions = {\n",
    "        \"+\": {\"args\": 2, \"f\": lambda x, y: x + y},\n",
    "        \"-\": {\"args\": 2, \"f\": lambda x, y: x - y},\n",
    "        \"*\": {\"args\": 2, \"f\": lambda x, y: x * y},\n",
    "        \"/\": {\"args\": 2, \"f\": lambda x, y: x / y if y != 0 else np.nan},\n",
    "        \"L\": {\"args\": 1, \"f\": lambda x: np.log(x) if x > 0 else np.nan},\n",
    "        \"E\": {\"args\": 1, \"f\": lambda x: np.exp(x)},\n",
    "        #\"K\": {\"args\": 1, \"f\": lambda x: np.log10(x) if x > 0 else np.nan},\n",
    "        #\"~\": {\"args\": 1, \"f\": lambda x: np.power(10, x)},\n",
    "        #\"S\": {\"args\": 1, \"f\": lambda x: np.sin(x)},\n",
    "        #\"C\": {\"args\": 1, \"f\": lambda x: np.cos(x)}\n",
    "    }\n",
    "    Chromosome.fitness_cases = [\n",
    "        ({\"a\": a}, GeneExpressionProgram.OBJECTIVE_FUNCTION(a))\n",
    "        for a in np.random.rand(GeneExpressionProgram.NUM_FITNESS_CASES) * (\n",
    "        GeneExpressionProgram.OBJECTIVE_MAX - GeneExpressionProgram.OBJECTIVE_MIN) + GeneExpressionProgram.OBJECTIVE_MIN\n",
    "    ]\n",
    "\n",
    "    Chromosome.linking_function = \"+\"\n",
    "    Chromosome.length = 60\n",
    "    Chromosome.head_length = 20\n",
    "    Chromosome.max_fitness = 1 #GeneExpressionProgram.FUNCTION_Y_RANGE * GeneExpressionProgram.NUM_FITNESS_CASES\n",
    "\n",
    "    average_fitnesses = []\n",
    "    best_fitnesses = []\n",
    "    answers = []\n",
    "\n",
    "    ans = None\n",
    "    for rep in range(GeneExpressionProgram.NUM_RUNS):\n",
    "        print(\"====================================== Rep %d ======================================\" % rep)\n",
    "        ans, gen_average_fitnesses, gen_best_fitnesses = GeneExpressionProgram.evolve()\n",
    "        average_fitnesses.append(gen_average_fitnesses)\n",
    "        best_fitnesses.append(gen_best_fitnesses)\n",
    "        ans.print_tree()\n",
    "        print(ans.genes)\n",
    "        answers.append(deepcopy(ans))\n",
    "\n",
    "    print(\"====================================== Random Search ======================================\")\n",
    "    random_search_individual, random_search_avg, random_search_best = GeneExpressionProgram.random_search(\n",
    "        GeneExpressionProgram.NUM_GENERATIONS,\n",
    "        GeneExpressionProgram.FITNESS_FUNCTION,\n",
    "        GeneExpressionProgram.FITNESS_FUNCTION_ARGS\n",
    "    )\n",
    "\n",
    "    # Good genes:\n",
    "    #ans = Chromosome(['+/aE?aa?aaEa*Ea+E/?L?aa??aaaaa???aaaaa???aa?a??a?aa???aaa??a?a??a?aaaaaa?aaa??aa', '?a???L*Laa+++L-?///aaa??????a?aa?aaa?aaa?aaa?aa??a?a?a?aa??aaa?a?aaaaaa?aa???aaa', '+L**E*aaEaa+E??*/a*-???a?aa?aaaa?aaaaaaa?a?aaaa?aa?aaaaaaa???aa?aaaa?aa??aaa????'])\n",
    "    #ans.ephemeral_random_constants = [0.15711028370174618, 0.8582913293794245, 0.08834476189713181, -0.6507871955790874, -0.5001482762755984, -0.36339831251011967, 0.49239731282122023, -0.7707384044151064, 0.33957437653782097, -0.050909069821311936, 0.40688394042469067, 0.8838615430620207, -0.03950637280673064, -0.8398663459127116, -0.9701111175669401, 0.16516078130630563, -0.5163031755060181, 0.26930803528455916, -0.7833159749333989, -0.7075160969083776, -0.6751546227334948, 0.1505636368911023, -0.16805240822390255, -0.34424168190370485, -0.8544980338428079, -0.01217703484745547, -0.24005751860391533, -0.5077198074421936, -0.47443544577074226, 0.5247967085947582, -0.22543318048008576, 0.4938865002308659, -0.6465093618701077, -0.19098460727467326, -0.2944062401077585, 0.7016839377380519, -0.14341637591100542, 0.23227088210671476, 0.36051772215302, 0.6509343605188611, -0.332150327502315, 0.3171544096208032, 0.2676850827993431, -0.46506262502073414, 0.843996276413379, -0.5614960005334659, 0.47891344757490373, -0.5575325624206526, -0.8156525364269045, -0.31652263727243746, 0.06884531540832572, 0.8836222097723032, 0.6601557412383419, -0.7869161076217597, -0.16110865846423783, -0.06702662866877018, -0.22568995470545228, 0.9438977429740325, -0.6410133183089668, 0.045353882523733624, 0.16786587502585948, -0.9008872395302681, 0.004355424045595413, 0.9179463218466064, 0.4105708444235643, -0.001799675111191501, -0.4201794697378056, -0.37672122028632216, 0.5906938113771141, 0.6004718433032417, -0.4698494772153414, 0.04238505345795085, 0.1657146428146341, 0.5148585050576182, 0.6955837854892111, -0.08812485635975653, -0.4101965485774235, 0.7234363214796748, 0.14285945798729927, 0.6450352601522822]\n",
    "    #ans.plot_solution(GeneExpressionProgram.OBJECTIVE_FUNCTION, GeneExpressionProgram.OBJECTIVE_MIN, GeneExpressionProgram.OBJECTIVE_MAX, [], [], \"a\")\n",
    "\n",
    "    #ans = Chromosome([0.10325095527275208, -0.6353207023355358, 0.3087002846077995, -0.646063772861873, -0.7604999404652737, -0.7026365243175745, -0.8546384172276964, -0.9210656698793773, -0.92473371383499159, 0.38475054329790037, 0.011392225394319944, -0.18840805047095377, 0.40103873272978197, 0.06328742279657229, 0.7176441901359412, -0.26941389839428265, 0.8330939983953638, -0.3291581824265808, -0.6081511326274629, 0.9437496769675557, 0.14649019501924365, -0.026586568971478375, 0.6942711369955727, -0.4952751886918654, 0.49348107956565457, 0.6373366861063929, -0.4180513467203788, 0.37222546291909575, 0.09407582333744235, -0.7976697480943422, -0.04875126785507633, 0.0738340898048051, 0.4574594498229325, 0.6176526203069368, -0.8316660105318336, 0.328772037314927, 0.854958551073552, -0.3935592840377773, -0.1523385500807699, -0.8272148377421589, 0.8580710774374207, -0.18672644199209687, 0.5393722078778671, -0.6906345459192962, -0.22443846331305117, 0.34512085908760115, 0.21516058072431332, -0.8499353974445536, 0.26989917691422316, -0.44201601204751584, 0.00887253706744362, -0.7836088648008288, -0.8679234491746306, -0.07209104779770104, 0.16000238129327515, 0.8890857194001152, -0.6325595149575627, -0.11804832230841544, -0.018168330413408817, -0.6561487673112327, 0.18667288624065526, -0.21171583904469493, -0.7074160949096726, -0.5647179047402893, 0.6466522462832285, 0.17157152187819658, -0.55051406791476754, -0.32459344245456756, -0.73568042741430184, 0.39543504084587644, -0.9332865781958235, 0.9903268434472272, 0.05041449572068979, 0.0798497529910911, -0.66617358746171074, 0.5071192429292173, 0.38014218205802264, -0.929868126429702, -0.23615736970440504, 0.5939936181262968])\n",
    "    #ans.ephemeral_random_constants = ['+EEEaa/L-+L-EL-L?aLEaa??????aa?aa?????aa??a???aaaaaaa????aa?aa??aa??a?a??aaa?aaa', '+EEE*aa?+?*/LLa/*+*-a?a?aaaa??a?aa?a??aa???aaaaa????a???????a?aaa????????aaaaaa?', '-L***aaa--+*+*+?+?--?aa??aa?aa?aa??a??a?aa??aaaaa?a??a?aa?aa?????aaaa??aa?aaa?a?']\n",
    "\n",
    "    best = max(answers, key=lambda c:c.fitness())\n",
    "    ans.plot_solution(GeneExpressionProgram.OBJECTIVE_FUNCTION, GeneExpressionProgram.OBJECTIVE_MIN, GeneExpressionProgram.OBJECTIVE_MAX, average_fitnesses[0], best_fitnesses[0], \"a\")\n",
    "    GeneExpressionProgram.plot_reps(average_fitnesses, best_fitnesses, random_search_avg, random_search_best)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cart_pole_real()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
